File: internal/app/approval_profile_test.go
package app

import "testing"

func TestApprovalProfileDangerous(t *testing.T) {
	if approvalProfileDangerous("workspace.read") || approvalProfileDangerous("web.search") {
		t.Fatal("low-risk tools must not be dangerous")
	}
	for _, name := range []string{"command.run", "desktop.open", "desktop.type", "cc.mouse_click", "im.send", "computer.act"} {
		if !approvalProfileDangerous(name) {
			t.Fatalf("%s must be dangerous", name)
		}
	}
	if companionToolPreapproved("command.run", false, false) || companionToolPreapproved("user.ask", false, false) {
		t.Fatal("dangerous and user.ask must not auto-approve")
	}
	if !companionToolPreapproved("workspace.read", true, false) {
		t.Fatal("low-risk companion tools stay preapproved")
	}
	if companionToolPreapproved("workspace.write", true, false) || companionToolPreapproved("workspace.edit", true, false) {
		t.Fatal("full-disk workspace writes must confirm once")
	}
	if !companionToolPreapproved("workspace.write", false, false) {
		t.Fatal("sandbox workspace writes stay preapproved")
	}
	// Standing computer-control enable pre-authorizes launch-shaped desktop
	// tools so a voice turn does not stall on an un-tappable approval.
	if companionToolPreapproved("desktop.open", false, false) {
		t.Fatal("desktop.open must confirm when computer control is off")
	}
	if !companionToolPreapproved("desktop.open", false, true) {
		t.Fatal("desktop.open must be preapproved once computer control is on")
	}
	if !companionToolPreapproved("media.play", false, true) {
		t.Fatal("media.play must be preapproved once computer control is on")
	}
	// The truly sensitive surface stays gated even with CC enabled.
	if companionToolPreapproved("desktop.type", false, true) || companionToolPreapproved("cc.mouse_click", false, true) || companionToolPreapproved("computer.act", false, true) {
		t.Fatal("pixel/keyboard desktop tools must still confirm even with CC on")
	}
}

File: web/src/session/approvalProfile.test.ts
import { expect, it } from 'vitest'
import { companionMayAutoApprove, isDangerousCompanionTool, isFullDiskCompanionWrite, sessionMayAutoApprove } from './approvalProfile'

it('keeps low-risk companion tools auto and dangerous ones confirmed', () => {
  expect(isDangerousCompanionTool('command.run')).toBe(true)
  expect(isDangerousCompanionTool('run_terminal_cmd')).toBe(true)
  expect(companionMayAutoApprove('run_terminal_cmd')).toBe(false)
  expect(isDangerousCompanionTool('cc.mouse_click')).toBe(true)
  expect(isDangerousCompanionTool('workspace.read')).toBe(false)
  expect(companionMayAutoApprove('workspace.read')).toBe(true)
  expect(companionMayAutoApprove('command.run')).toBe(false)
  expect(companionMayAutoApprove('user.ask')).toBe(false)
  expect(isFullDiskCompanionWrite('workspace.write')).toBe(true)
  expect(companionMayAutoApprove('workspace.write', false)).toBe(true)
  expect(companionMayAutoApprove('workspace.write', true)).toBe(false)
  expect(companionMayAutoApprove('workspace.edit')).toBe(false)
})

it('lets a typed full-access turn skip a second approval except user.ask', () => {
  expect(sessionMayAutoApprove('office.generate')).toBe(true)
  expect(sessionMayAutoApprove('docx.gen')).toBe(true)
  expect(sessionMayAutoApprove('workspace.write')).toBe(true)
  expect(sessionMayAutoApprove('command.run')).toBe(true)
  expect(sessionMayAutoApprove('skill.try')).toBe(true)
  expect(sessionMayAutoApprove('user.ask')).toBe(false)
})
